--- a/extra/Makefile.static	2016-01-02 18:00:30.000000000 +0000
+++ b/extra/Makefile.static	2016-06-17 10:21:54.545375749 +0000
@@ -1,17 +1,14 @@
 # Simple makefile to build inih as a static library using g++
 
-SRC = ../ini.c
-OBJ = $(SRC:.c=.o)
+SRC = ../ini.c ../cpp/INIReader.cpp
+OBJ := $(patsubst %.c,%.o,$(SRC))
+OBJ := $(patsubst %.cpp,%.o,$(OBJ))
 OUT = libinih.a
 INCLUDES = -I..
-CCFLAGS = -g -O2
-CC = g++
+CCFLAGS = -g -O2 -fPIC $(INCLUDES)
 
 default: $(OUT)
 
-.c.o:
-	$(CC) $(INCLUDES) $(CCFLAGS) $(EXTRACCFLAGS) -c $< -o $@
-
 $(OUT): $(OBJ)
 	ar rcs $(OUT) $(OBJ) $(EXTRAARFLAGS)
 
